File: frontend/src/api/assignmentStatus.ts
/**
 * 作业生命周期状态的中文标签。
 *
 * 后端 `backend/.../assignment/AssignmentStatus.java` 是唯一来源，共 7 项。
 * 标签表、已知集合与类型都从下面这一份 `ASSIGNMENT_STATUS_OPTIONS` 派生，
 * 不再手写第二份对象——「后端加了状态、前端漏补一项」这类漂移都源于两处清单。
 *
 * 注意历史取值 `IMPORTED`：旧版教师端 Excel 导入答案时写入过它，后端读取时会映射成
 * `SUBMITTING`，所以前端只可能收到这 7 个取值之一，不需要为它留标签。
 */
export const ASSIGNMENT_STATUS_OPTIONS = [
  { value: 'DRAFT', label: '草稿' },
  { value: 'OCR_REVIEW', label: '待校对试卷' },
  { value: 'PUBLISHED', label: '已发布' },
  { value: 'SUBMITTING', label: '收卷中' },
  { value: 'GRADING', label: '批改中' },
  { value: 'REVIEWING', label: '待复核' },
  { value: 'COMPLETED', label: '已完成' },
] as const

export type AssignmentStatus = typeof ASSIGNMENT_STATUS_OPTIONS[number]['value']

const ASSIGNMENT_STATUS_LABELS: Record<string, string> = Object.fromEntries(
  ASSIGNMENT_STATUS_OPTIONS.map(option => [option.value, option.label]),
)

/**
 * 认不出的取值原样返回。
 *
 * 形参刻意是 `string` 而不是 `AssignmentStatus`：服务端可能比前端先加上新状态，
 * 显示英文总比显示"未知"好——既不丢信息，也让新取值在界面上立刻看得出来。
 */
export function assignmentStatusLabel(status: string): string {
  return ASSIGNMENT_STATUS_LABELS[status] ?? status
}

/** 已发布之后的状态都不允许教师再改动题目清单，界面据此禁用编辑入口。 */
export function isPublishedStatus(status: string): boolean {
  return status !== 'DRAFT' && status !== 'OCR_REVIEW'
}

const SUBMISSION_STATUS_LABELS: Record<string, string> = {
  NOT_SUBMITTED: '未提交',
  // 旧版教师端导入的答案，学生端只当作"已有答卷"展示。
  IMPORTED: '已提交',
}

/**
 * 学生自己的提交摘要标签。
 *
 * 目前只有后端在本阶段能产出的两个取值；提交版本状态（草稿、处理中、待校对、
 * 已退回等）属于后续任务的提交版本模型，届时在这里补齐而不是先写一批用不到的值。
 */
export function submissionStatusLabel(status: string): string {
  return SUBMISSION_STATUS_LABELS[status] ?? status
}

/** 已知取值集合，供测试核对后端产出是否都在标签表里。 */
export const KNOWN_ASSIGNMENT_STATUSES: string[] = ASSIGNMENT_STATUS_OPTIONS.map(option => option.value)

File: backend/src/main/java/com/homework/analysis/assignment/AssignmentStatus.java
package com.homework.analysis.assignment;

import java.util.EnumSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * 作业生命周期状态。
 *
 * <p>状态只能沿一条线向前走：{@code DRAFT/OCR_REVIEW → PUBLISHED → SUBMITTING → GRADING
 * → REVIEWING → COMPLETED}。刻意不提供 {@code canTransitionTo} 之外的“任意设置状态”入口，
 * 否则迟早会有人从接口层直接把作业改成 {@code COMPLETED}，跳过评分与复核。
 *
 * <p>唯一允许的向后动作是“教师退回”，它作用在提交版本上而不是作业状态上，
 * 因此不在本枚举里开口子。
 */
public enum AssignmentStatus {
    DRAFT,
    OCR_REVIEW,
    PUBLISHED,
    SUBMITTING,
    GRADING,
    REVIEWING,
    COMPLETED;

    /**
     * 历史取值。
     *
     * <p>教师端 Excel 导入答案属于旧链路，它写入的 {@code IMPORTED} 表示“已有答案待批改”，
     * 语义上等价于 {@code SUBMITTING}。迁移前的正式库与演示库里都有这种行，
     * 直接按未知状态抛错会让历史作业全部打不开，所以在这里映射一次。
     */
    private static final Map<String, AssignmentStatus> LEGACY_NAMES = Map.of("IMPORTED", SUBMITTING);

    private static final Map<AssignmentStatus, Set<AssignmentStatus>> ALLOWED_NEXT = Map.of(
        DRAFT, EnumSet.of(PUBLISHED),
        OCR_REVIEW, EnumSet.of(PUBLISHED),
        PUBLISHED, EnumSet.of(SUBMITTING),
        SUBMITTING, EnumSet.of(GRADING),
        GRADING, EnumSet.of(REVIEWING),
        REVIEWING, EnumSet.of(COMPLETED),
        COMPLETED, EnumSet.noneOf(AssignmentStatus.class));

    public static AssignmentStatus parse(String raw) {
        if (raw == null) {
            throw new IllegalArgumentException("作业状态为空");
        }
        String normalized = raw.trim().toUpperCase(Locale.ROOT);
        AssignmentStatus legacy = LEGACY_NAMES.get(normalized);
        if (legacy != null) {
            return legacy;
        }
        try {
            return valueOf(normalized);
        } catch (IllegalArgumentException exception) {
            // 库里出现无法识别的状态属于数据损坏，不是用户错误：让它在 500 里暴露出来，
            // 而不是悄悄退回某个默认状态继续跑。
            throw new IllegalArgumentException("无法识别的作业状态：" + raw, exception);
        }
    }

    public boolean canTransitionTo(AssignmentStatus next) {
        return ALLOWED_NEXT.getOrDefault(this, Set.of()).contains(next);
    }

    /**
     * 是否已进入学生可见范围。
     *
     * <p>仅作兜底判断：学生查询以 {@code assignment.published_at is not null} 为准，
     * 因为“已发布”是一次性的时间事实，不会被后续状态推进抹掉。
     */
    public boolean visibleToStudents() {
        return this != DRAFT && this != OCR_REVIEW;
    }
}
